From f3bb406f85a6e04f7ae1ef6df25d5ae7c006eda7 Mon Sep 17 00:00:00 2001
From: Robert Dabrowski <robertx.dabrowski@intel.com>
Date: Mon, 28 Sep 2020 10:51:25 +0200
Subject: [PATCH] PONRTSYS-7969: Configure LCT traffic classification rules -
 Remove LCT TRAP rules from QoS driver. Configure them from OMCI ME   level
 instead

---
 drivers/net/datapath/pon_qos/pon_qos_tc_lct.c | 213 --------------------------
 1 file changed, 213 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_lct.c b/drivers/net/datapath/pon_qos/pon_qos_tc_lct.c
index 4c22142fede4..60738b2c6622 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_lct.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_lct.c
@@ -48,215 +48,6 @@ static struct pon_qos_lct_control pon_qos_lct_db[PON_QOS_LCT_MAX] = {
 	 true},
 };
 
-static int pce_rule_lct_mac_to_cpu(struct net_device *dev,
-				   struct core_ops *gsw_handle,
-				   dp_subif_t dp_subif,
-				   uint32_t idx)
-{
-	GSW_PCE_rule_t pce_rule = {0};
-	GSW_return_t ret;
-	s32 i;
-
-	pce_rule.pattern.nIndex = idx;
-	pce_rule.pattern.bEnable = 1;
-	pce_rule.pattern.bPortIdEnable = 1;
-	pce_rule.pattern.nPortId = dp_subif.port_id;
-	pce_rule.pattern.bSubIfIdEnable = 1;
-	pce_rule.pattern.nSubIfId = dp_subif.subif;
-	pce_rule.pattern.bMAC_DstEnable = 1;
-	for (i = 0; i < ARRAY_SIZE(pce_rule.pattern.nMAC_Dst); i++)
-		pce_rule.pattern.nMAC_Dst[i] = dev->dev_addr[i];
-
-	pce_rule.action.ePortMapAction = GSW_PCE_ACTION_PORTMAP_CPU;
-	pce_rule.action.nTrafficClassAlternate = 11;
-
-	ret = gsw_handle->gsw_tflow_ops.TFLOW_PceRuleWrite(gsw_handle,
-							   &pce_rule);
-	if (ret != GSW_statusOk) {
-		pr_err("%s: TFLOW_PceRuleWrite [LCT MAC to CPU] for interface \"%s\" failed: %d\n",
-		       __func__, dev->name, ret);
-		return ret;
-	}
-
-	return 0;
-}
-
-static int pce_rule_lct_mc_bc_to_cpu(struct net_device *dev,
-				     struct core_ops *gsw_handle,
-				     dp_subif_t dp_subif,
-				     uint32_t idx)
-{
-	GSW_PCE_rule_t pce_rule = {0};
-	GSW_return_t ret;
-
-	pce_rule.pattern.nIndex = idx;
-	pce_rule.pattern.bEnable = 1;
-	pce_rule.pattern.bPortIdEnable = 1;
-	pce_rule.pattern.nPortId = dp_subif.port_id;
-	pce_rule.pattern.bSubIfIdEnable = 1;
-	pce_rule.pattern.nSubIfId = dp_subif.subif;
-	pce_rule.pattern.bMAC_DstEnable = 1;
-	/* IPv6 multicast: 33:33:xx:xx:xx:xx.
-	 * Each DstMask bit represents a nibble of the MAC as do-not-care.
-	 */
-	pce_rule.pattern.nMAC_Dst[0] = 0x33;
-	pce_rule.pattern.nMAC_Dst[1] = 0x33;
-	pce_rule.pattern.nMAC_DstMask = 0xFF;
-	pce_rule.pattern.bEtherTypeEnable = 1;
-	pce_rule.pattern.nEtherType = ETH_P_IPV6;
-
-	pce_rule.action.ePortMapAction = GSW_PCE_ACTION_PORTMAP_CPU;
-	pce_rule.action.nTrafficClassAlternate = 12;
-
-	ret = gsw_handle->gsw_tflow_ops.TFLOW_PceRuleWrite(gsw_handle,
-							   &pce_rule);
-	if (ret != GSW_statusOk) {
-		pr_err("%s: TFLOW_PceRuleWrite [LCT Multicast Broadcast to CPU] for interface \"%s\" failed: %d\n",
-		       __func__, dev->name, ret);
-		return ret;
-	}
-
-	return 0;
-}
-
-static int pce_rule_lct_arp_bc_to_cpu(struct net_device *dev,
-				      struct core_ops *gsw_handle,
-				      dp_subif_t dp_subif,
-				      uint32_t idx)
-{
-	GSW_PCE_rule_t pce_rule = {0};
-	GSW_return_t ret;
-	s32 i;
-
-	pce_rule.pattern.nIndex = idx;
-	pce_rule.pattern.bEnable = 1;
-	pce_rule.pattern.bPortIdEnable = 1;
-	pce_rule.pattern.nPortId = dp_subif.port_id;
-	pce_rule.pattern.bSubIfIdEnable = 1;
-	pce_rule.pattern.nSubIfId = dp_subif.subif;
-	pce_rule.pattern.bMAC_DstEnable = 1;
-	for (i = 0; i < 6; i++)
-		pce_rule.pattern.nMAC_Dst[i] = 0xFF;
-	pce_rule.pattern.bEtherTypeEnable = 1;
-	pce_rule.pattern.nEtherType = ETH_P_ARP;
-
-	pce_rule.action.ePortMapAction = GSW_PCE_ACTION_PORTMAP_CPU;
-	pce_rule.action.nTrafficClassAlternate = 13;
-
-	ret = gsw_handle->gsw_tflow_ops.TFLOW_PceRuleWrite(gsw_handle,
-							   &pce_rule);
-	if (ret != GSW_statusOk) {
-		pr_err("%s: TFLOW_PceRuleWrite [LCT ARP Broadcast to CPU] for interface \"%s\" failed: %d\n",
-		       __func__, dev->name, ret);
-		return ret;
-	}
-
-	return 0;
-}
-
-static int pce_rule_lct_ptp_to_cpu(struct net_device *dev,
-				   struct core_ops *gsw_handle,
-				   dp_subif_t dp_subif,
-				   uint32_t idx)
-{
-	GSW_PCE_rule_t pce_rule = {0};
-	GSW_return_t ret;
-
-	pce_rule.pattern.nIndex = idx;
-	pce_rule.pattern.bEnable = 1;
-	pce_rule.pattern.bPortIdEnable = 1;
-	pce_rule.pattern.nPortId = dp_subif.port_id;
-	pce_rule.pattern.bSubIfIdEnable = 1;
-	pce_rule.pattern.nSubIfId = dp_subif.subif;
-	pce_rule.pattern.bEtherTypeEnable = 1;
-	pce_rule.pattern.nEtherType = ETH_P_1588;
-
-	pce_rule.action.ePortMapAction = GSW_PCE_ACTION_PORTMAP_CPU;
-	pce_rule.action.eTrafficClassAction = 2;
-	/* Use high priroity queue for PTP next to OMCI */
-	pce_rule.action.nTrafficClassAlternate = 9;
-
-	ret = gsw_handle->gsw_tflow_ops.TFLOW_PceRuleWrite(gsw_handle,
-							   &pce_rule);
-	if (ret != GSW_statusOk) {
-		pr_err("%s: TFLOW_PceRuleWrite [PTP to CPU] for interface \"%s\" failed: %d\n",
-		       __func__, dev->name, ret);
-		return ret;
-	}
-
-	return 0;
-}
-
-static void pce_rule_remove(uint32_t idx)
-{
-	struct core_ops *gsw_handle;
-	GSW_PCE_ruleDelete_t pce_rule = {0};
-	GSW_return_t ret;
-
-	gsw_handle = gsw_get_swcore_ops(0);
-	if (!gsw_handle)
-		return;
-
-	pce_rule.nIndex = idx;
-
-	ret = gsw_handle->gsw_tflow_ops.TFLOW_PceRuleDelete(gsw_handle,
-							    &pce_rule);
-	if (ret != GSW_statusOk)
-		pr_err("%s: TFLOW_PceRuleDelete with idx \"%u\" failed: %d\n",
-		       __func__, idx, ret);
-}
-
-static int pce_rules_lct_to_cpu(struct net_device *dev,
-				uint8_t lct_idx)
-{
-	struct core_ops *gsw_handle;
-	dp_subif_t dp_subif;
-	s32 ret;
-
-	if (pon_qos_lct_db[lct_idx].configured == PON_QOS_PCE_CONFIGURED)
-		return 0;
-
-	gsw_handle = gsw_get_swcore_ops(0);
-	if (!gsw_handle)
-		return -EINVAL;
-
-	ret = dp_get_netif_subifid(dev, NULL, NULL, NULL, &dp_subif, 0);
-	if (ret != DP_SUCCESS)
-		return ENODEV;
-
-	ret = pce_rule_lct_mac_to_cpu(dev, gsw_handle, dp_subif,
-				      pon_qos_lct_db[lct_idx].mac_rule_idx);
-	if (ret != 0)
-		return ret;
-
-	ret = pce_rule_lct_mc_bc_to_cpu(dev, gsw_handle, dp_subif,
-					pon_qos_lct_db[lct_idx].mc_rule_idx);
-	if (ret != 0) {
-		pce_rule_remove(pon_qos_lct_db[lct_idx].mac_rule_idx);
-		return ret;
-	}
-
-	ret = pce_rule_lct_arp_bc_to_cpu(dev, gsw_handle, dp_subif,
-					 pon_qos_lct_db[lct_idx].arp_rule_idx);
-	if (ret != 0) {
-		pce_rule_remove(pon_qos_lct_db[lct_idx].mac_rule_idx);
-		pce_rule_remove(pon_qos_lct_db[lct_idx].mc_rule_idx);
-		return ret;
-	}
-
-	ret = pce_rule_lct_ptp_to_cpu(dev, gsw_handle, dp_subif,
-				      pon_qos_lct_db[lct_idx].ptp_rule_idx);
-	if (ret != 0) {
-		pce_rule_remove(pon_qos_lct_db[lct_idx].mac_rule_idx);
-		pce_rule_remove(pon_qos_lct_db[lct_idx].mc_rule_idx);
-		pce_rule_remove(pon_qos_lct_db[lct_idx].arp_rule_idx);
-		return ret;
-	}
-
-	pon_qos_lct_db[lct_idx].configured = PON_QOS_PCE_CONFIGURED;
-	return 0;
-}
-
 /* TODO: Remove this function when DP is fixed */
 static int cpu_port_disconnect(struct net_device *dev)
 {
@@ -326,10 +117,6 @@ int pon_qos_lct_configure(struct net_device *dev,
 		pon_qos_lct_db[i].cpu_port_mapped = false;
 	}
 
-	ret = pce_rules_lct_to_cpu(lct_dev, i);
-	if (ret != 0)
-		return ret;
-
 	rule->act.ract.act = DP_BP_REASSIGN;
 	rule->act.ract.bp_dev = lct_dev;
 
